migrations/000024_cpu_usage.up.sql

```diff
@@ -0,0 +1,58 @@
+BEGIN;
+
+SET search_path = public, pg_catalog;
+
+CREATE TABLE IF NOT EXISTS cpu_usage_event_types (
+    id uuid NOT NULL DEFAULT uuid_generate_v4(),
+    name character varying(32) NOT NULL,
+    description text,
+    last_modified timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+
+    PRIMARY KEY(id)
+);
+
+DROP TRIGGER IF EXISTS cpu_usage_event_types_last_modified_trigger ON cpu_usage_event_types CASCADE;
+CREATE TRIGGER cpu_usage_event_types_last_modified_trigger
+    BEFORE UPDATE ON cpu_usage_event_types
+    FOR EACH ROW
+    EXECUTE PROCEDURE moddatetime (last_modified);
+
+CREATE TABLE IF NOT EXISTS cpu_usage_events (
+    id uuid NOT NULL DEFAULT uuid_generate_v4(),
+    record_date timestamp NOT NULL,
+    effective_date timestamp NOT NULL,
+    event_type_id uuid NOT NULL,
+    value bigint NOT NULL,
+    created_by uuid NOT NULL,
+    last_modified timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+
+    FOREIGN KEY (event_type_id) REFERENCES cpu_usage_event_types(id),
+    FOREIGN KEY (created_by) REFERENCES users(id),
+    PRIMARY KEY (id)
+);
+
+DROP TRIGGER IF EXISTS cpu_usage_events_last_modified_trigger ON cpu_usage_events CASCADE;
+CREATE TRIGGER cpu_usage_events_last_modified_trigger
+    BEFORE UPDATE ON cpu_usage_events
+    FOR EACH ROW
+    EXECUTE PROCEDURE moddatetime (last_modified);
+
+CREATE TABLE IF NOT EXISTS cpu_usage_totals (
+    id uuid NOT NULL DEFAULT uuid_generate_v4(),
+    user_id uuid NOT NULL,
+    total bigint NOT NULL DEFAULT 0,
+    effective_start_date timestamp NOT NULL,
+    effective_end_date timestamp NOT NULL,
```

Comment:
Might be the same question here, but also since it's a totals table, what happens if there are overlapping date ranges in this table?

Reply:
This might need a uniqueness constraint added as you mentioned, on the user_id, effective_start_date, and effective_end_date, Maybe another trigger, too, to prevent inserts with an effective_(start|end)_date that falls between an existing effective_(start|end)_date range for the same user.

Reply:
I changed this to be a tsrange with an exclusion constraint that prevents a user from having overlapping ranges. It requires adding the btree_gist extension.